### DISCUSSION TITLE
Why ToCanonical is not invertible ?

### DISCUSSION THREAD

**Discussion by lchauvin** [TorchIO-project/torchio#1139] [Q&A]
Hello,

I was wondering why the transform ToCanonical is not invertible ?

I'd like to put all images in the same space for training and inference, but I'd like to have them back in their original space after, to save them.
However, it's not possible because of the ToCanonical transform, that is not invertible, and so using 'apply_invert_transform' doesn't put them back in their original space.

Why would ToCanonical not be invertible ?

Thank you.

**Comment by romainVala** [comment]
Hello
that's a good question, I agree this transform should be invertible, but this has not been implemented yet.
the workaround I use is the following:

```
ImagePath = 'path_to_the_nifti_volume'
volume = torchio.ScalarImage(path= ImagePath)
tstd = torchio.ToCanonical()
volume = tstd(volume)

tresamp = torchio.Resample(target = ImagePath ,image_interpolation='nearest' )

volume_original_space = tresamp(volume)

```
Note that `image_interpolation='nearest' `is to avoid any interpolation so that you get exactly the initial volume

**Comment by fepegar** [comment]
I guess there are a couple of historical reasons.

1. `ToCanonical` only reorients from some orientation to RAS+. We currently don't have a transform to reorient to anything else. If you are interested, feel free to contribute!
1. When we store the transforms history, we store transforms parameters, not image attributes (which in this case would be necessary to reorient to the original orientation). Again, I don't think there are good arguments against this and could be implemented.

I typically use the same solution as @romainVala's.

**Reply by fepegar** [reply]
Note: @romainVala, I think the interpolation choice shouldn't matter as no actual resampling is needed here, right? Maybe `'nearest'` would just be faster.

**Reply by romainVala** [reply]
Hi
yes you are right, I am always suspicious with resampling, 
I just test and with default (linear) interpolation, I end up with small numerical difference for  (tresamp(tstd(i1)) -i1)  max error is inferior to 10 exp -11 (good enough!) but if I use nearest, there are no differences at all (even better !) ...

@fepegar  in the case of the ToCanonical transform affine image attribute is also the transforms attribute (implicitly used by nibabel) and the one we need to make the way back. so a good argument for it

regarding a reorient transform, no sure we need it since it is much easier to make it with resampling. 

this being say, it would be a nice random transform to add (random reorient (ie just mixing orientation without resampling). I may work on it later (as it may be usefull for orientation agnostic network)

Cheers
